File: bindings/python/test_zcio.py
"""Smoke tests for the zcio Python binding. Run: python test_zcio.py
Requires the shared lib built at ../../build (cmake -DZCIO_BUILD_SHARED=ON)."""
import os
import socket
import sys
import threading
import time

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import zcio


class SkipTest(Exception):
    """Raise to mark a test as skipped (environment-dependent)."""


def _free_port():
    """Ask the OS for a currently-free TCP port. Avoids TIME_WAIT collisions
    between back-to-back runs (zcio's C server sets no SO_REUSEADDR)."""
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind(("127.0.0.1", 0))
    port = s.getsockname()[1]
    s.close()
    return port


# ---------------------------------------------------------------------------
# basics
# ---------------------------------------------------------------------------
def test_version():
    v = zcio.version()
    assert v and v[0].isdigit(), v
    print("version:", v)


def test_ipv4():
    assert zcio.is_ipv4("127.0.0.1")
    assert not zcio.is_ipv4("999.0.0.1")
    assert zcio.resolve_ipv4("8.8.8.8") == "8.8.8.8"


def test_dns_query_a():
    addrs = zcio.dns_query_a("localhost")
    assert isinstance(addrs, list) and len(addrs) >= 1, addrs
    # at least one should be a v4 literal (commonly 127.0.0.1)
    assert any(zcio.is_ipv4(a) for a in addrs), addrs


def test_dns_query_aaaa():
    # may legitimately be empty on hosts without IPv6; just must not crash.
    addrs = zcio.dns_query_aaaa("localhost")
    assert isinstance(addrs, list)


def test_local_ipv4():
    ip = zcio.local_ipv4()
    # returns a string or None; if a string it should look like a v4 address
    assert ip is None or zcio.is_ipv4(ip), ip


# ---------------------------------------------------------------------------
# ring
# ---------------------------------------------------------------------------
def test_ring():
    r = zcio.Ring(64)
    assert r.write(b"hello zcio") == 10
    assert r.available == 10
    assert r.available_write == 54
    assert r.read(10) == b"hello zcio"
    assert r.available == 0
    r.close()


# ---------------------------------------------------------------------------
# serial
# ---------------------------------------------------------------------------
def test_serial():
    s = zcio.Serial(256)
    s.write_i32(-42)
    s.write_u64(0xCAFEBABE)
    s.write_f64(2.5)
    s.write_str("hi")
    rd = zcio.Serial(256, buffer=s.buffer)
    assert rd.read_i32() == -42
    assert rd.read_u64() == 0xCAFEBABE
    assert rd.read_f64() == 2.5
    assert rd.read_str() == "hi"


def test_serial_all_scalars():
    s = zcio.Serial(512)
    s.write_i8(-5)
    s.write_u8(250)
    s.write_i16(-1000)
    s.write_u16(60000)
    s.write_i32(-100000)
    s.write_u32(4000000000)
    s.write_i64(-9_000_000_000)
    s.write_u64(18_000_000_000)
    s.write_f32(1.5)
    s.write_f64(3.25)
    rd = zcio.Serial(512, buffer=s.buffer)
    assert rd.read_i8() == -5
    assert rd.read_u8() == 250
    assert rd.read_i16() == -1000
    assert rd.read_u16() == 60000
    assert rd.read_i32() == -100000
    assert rd.read_u32() == 4000000000
    assert rd.read_i64() == -9_000_000_000
    assert rd.read_u64() == 18_000_000_000
    assert abs(rd.read_f32() - 1.5) < 1e-6
    assert rd.read_f64() == 3.25


def test_serial_byte_and_bytes():
    s = zcio.Serial(64)
    s.write_byte(0xAB)
    s.write_bytes(b"raw-bytes")
    rd = zcio.Serial(64, buffer=s.buffer)
    assert rd.read_byte() == 0xAB
    assert rd.read_bytes(9) == b"raw-bytes"


def test_serial_bits_roundtrip():
    pattern = [True, False, True, True, False, False, True, False, True]
    s = zcio.Serial(64, bit_stream=True)
    s.write_bits(pattern)
    s.synchronize()
    rd = zcio.Serial(64, bit_stream=True, buffer=s.buffer)
    got = rd.read_bits(len(pattern))
    assert got == pattern, (got, pattern)


def test_serial_single_bits():
    s = zcio.Serial(16, bit_stream=True)
    for b in (True, True, False, True, False):
        s.write_bit(b)
    s.synchronize()
    rd = zcio.Serial(16, bit_stream=True, buffer=s.buffer)
    assert [rd.read_bit() for _ in range(5)] == [True, True, False, True, False]


def test_serial_positioning():
    s = zcio.Serial(64)
    s.write_i32(1)
    s.write_i32(2)
    assert s.write_pos == 8
    s.set_write_pos(0)
    s.write_i32(99)
    rd = zcio.Serial(64, buffer=s.buffer)
    assert rd.read_i32() == 99
    assert rd.read_pos == 4
    rd.set_read_pos(4)
    assert rd.read_i32() == 2


def test_serial_count_mode():
    s = zcio.Serial(count=True)
    s.write_i32(0)        # 4
    s.write_f64(0.0)      # 8
    s.write_str("abc")    # 8 (u64 len) + 3
    assert s.write_pos == 4 + 8 + 8 + 3, s.write_pos


def test_serial_read_status():
    s = zcio.Serial(8)
    s.write_i16(7)
    rd = zcio.Serial(8, buffer=s.buffer)
    assert rd.read_i16() == 7
    assert not rd.short_read()
    rd.read_bytes(100)  # overread the 8-byte buffer
    assert rd.short_read() or rd.read_empty() or rd.read_eof()
    assert isinstance(rd.last_read(), int)
    rd.clear_read()


# ---------------------------------------------------------------------------
# streams
# ---------------------------------------------------------------------------
def test_memory_stream():
    m = zcio.MemoryStream(32)
    assert m.write(b"0123456789") == 10
    m.seek_read(0)
    assert m.read_full(10) == b"0123456789"
    m.close()


def test_stream_verbs():
    m = zcio.MemoryStream(32)
    m.write(b"abcdef")
    assert m.flush() == 0
    m.seek_read(0)
    assert m.available >= 0
    assert m.read(3) == b"abc"
    assert m.name  # backend identifier, e.g. "membuf"/"memory"
    m.close()


def test_copy_ring_to_memory():
    r = zcio.Ring(64)
    r.write(b"payload-via-copy")
    src = r.as_stream(take_ownership=True)
    dst = zcio.MemoryStream(64)
    n = zcio.copy(dst, src, limit=len(b"payload-via-copy"))
    assert n == 16, n
    dst.seek_read(0)
    assert dst.read_full(16) == b"payload-via-copy"
    src.close()
    dst.close()


# ---------------------------------------------------------------------------
# TCP
# ---------------------------------------------------------------------------
def test_tcp_loopback():
    port = _free_port()
    srv = zcio.TcpServer(port)
    cli = zcio.TcpClient("127.0.0.1", port)
    conn = srv.accept(2000)
    assert conn is not None
    cli.stream.write(b"ping")
    assert conn.read_full(4) == b"ping"
    cli.close()
    srv.close()


def test_tcp_wait_and_avail():
    port = _free_port()
    srv = zcio.TcpServer(port)
    cli = zcio.TcpClient("127.0.0.1", port)
    conn = srv.accept_conn(2000)
    assert conn is not None
    cli.stream.write(b"hello!")
    # server side should become readable
    assert conn.wait_readable(2000) >= 0
    # client should be writable
    assert cli.wait_writable(2000) >= 0
    data = conn.stream.read_full(6)
    assert data == b"hello!"
    # now have server reply so the client has data to read
    conn.stream.write(b"yo")
    assert cli.wait_readable(2000) >= 0
    _ = cli.bytes_available  # just exercise it (>= 0 or -1 acceptable)
    assert cli.stream.read_full(2) == b"yo"
    cli.close()
    srv.close()


def test_tcp_two_clients_and_close():
    port = _free_port()
    srv = zcio.TcpServer(port)
    c1 = zcio.TcpClient("127.0.0.1", port)
    conn1 = srv.accept_conn(2000)
    c2 = zcio.TcpClient("127.0.0.1", port)
    conn2 = srv.accept_conn(2000)
    assert conn1 is not None and conn2 is not None
    assert conn1.id != conn2.id, (conn1.id, conn2.id)
    c1.stream.write(b"aaa")
    c2.stream.write(b"bbb")
    assert conn1.stream.read_full(3) == b"aaa"
    assert conn2.stream.read_full(3) == b"bbb"
    # close just one client connection on the server
    assert srv.close_client(conn1.id) == 0
    c1.close()
    c2.close()
    srv.close()


# ---------------------------------------------------------------------------
# UDP
# ---------------------------------------------------------------------------
def test_udp_loopback():
    port = _free_port()
    srv = zcio.UdpServer(port)
    cli = zcio.UdpClient("127.0.0.1", port)
    cli.stream.write(b"udp-ping")
    pkt = srv.receive(timeout_us=2_000_000)
    assert pkt is not None
    assert pkt.read(8) == b"udp-ping"
    cli.close()
    srv.close()


def test_udp_two_clients():
    port = _free_port()
    srv = zcio.UdpServer(port)
    c1 = zcio.UdpClient("127.0.0.1", port)
    c2 = zcio.UdpClient("127.0.0.1", port)
    c1.stream.write(b"one")
    c2.stream.write(b"two")
    got = set()
    for _ in range(2):
        pkt = srv.receive(timeout_us=2_000_000)
        assert pkt is not None
        got.add(pkt.read(3))
    assert got == {b"one", b"two"}, got
    c1.close()
    c2.close()
    srv.close()


# ---------------------------------------------------------------------------
# Multicast (best-effort; SKIP if no datagram arrives)
# ---------------------------------------------------------------------------
def test_multicast_loopback():
    group, port = "239.255.0.1", _free_port()
    try:
        recv = zcio.McastReceiver(group, port)
    except zcio.ZcioError as e:
        raise SkipTest(f"mcast receiver open failed: {e}")
    try:
        send = zcio.McastSender(group, port)
    except zcio.ZcioError as e:
        recv.close()
        raise SkipTest(f"mcast sender open failed: {e}")
    try:
        # exercise open/stream/write/free regardless of routing
        send.stream.write(b"mcast-hi")
        time.sleep(0.1)
        send.stream.write(b"mcast-hi")
        # best-effort receive; many headless envs have no mcast loopback route
        data = recv.stream.read(8)
        if not data:
            raise SkipTest("no multicast datagram (no loopback route, headless)")
        assert data == b"mcast-hi", data
    finally:
        send.close()
        recv.close()


# ---------------------------------------------------------------------------
# HTTP (offline: tiny server in a thread)
# ---------------------------------------------------------------------------
def _start_http_server():
    """Raw-socket HTTP/1.1 server returning 200 OK with a fixed body.
    Returns (port, stop_fn)."""
    srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    srv.bind(("127.0.0.1", 0))
    srv.listen(8)
    port = srv.getsockname()[1]
    stop = threading.Event()

    def serve():
        srv.settimeout(0.25)
        while not stop.is_set():
            try:
                conn, _ = srv.accept()
            except socket.timeout:
                continue
            except OSError:
                break
            try:
                conn.settimeout(1.0)
                # drain request headers
                buf = b""
                while b"\r\n\r\n" not in buf:
                    chunk = conn.recv(4096)
                    if not chunk:
                        break
                    buf += chunk
                body = b"hello from test server"
                resp = (
                    b"HTTP/1.1 200 OK\r\n"
                    b"Content-Type: text/plain\r\n"
                    b"Content-Length: " + str(len(body)).encode() + b"\r\n"
                    b"Connection: close\r\n"
                    b"\r\n" + body
                )
                conn.sendall(resp)
            except OSError:
                pass
            finally:
                conn.close()

    t = threading.Thread(target=serve, daemon=True)
    t.start()

    def shutdown():
        stop.set()
        t.join(timeout=2)
        srv.close()

    return port, shutdown


def test_http_get_offline():
    port, shutdown = _start_http_server()
    try:
        time.sleep(0.05)
        resp = zcio.http_get(f"http://127.0.0.1:{port}/")
        assert resp["status"] == 200, resp
        assert b"hello from test server" in resp["body"], resp["body"]
    finally:
        shutdown()


def test_http_request_verb():
    port, shutdown = _start_http_server()
    try:
        time.sleep(0.05)
        resp = zcio.http_request("GET", f"http://127.0.0.1:{port}/",
                                 headers={"X-Test": "1"})
        assert resp["status"] == 200, resp
        assert b"hello from test server" in resp["body"]
    finally:
        shutdown()


def test_http_bad_host():
    # A clearly-unresolvable host: status 0, no crash.
    resp = zcio.http_get("http://no-such-host.invalid.zzz/")
    assert resp["status"] == 0, resp


# ---------------------------------------------------------------------------
# runner
# ---------------------------------------------------------------------------
def main():
    tests = [v for k, v in sorted(globals().items()) if k.startswith("test_")]
    failed = skipped = 0
    for t in tests:
        try:
            t()
            print(f"[ ok ] {t.__name__}")
        except SkipTest as e:
            skipped += 1
            print(f"[SKIP] {t.__name__}: {e}")
        except Exception as e:  # noqa: BLE001
            failed += 1
            print(f"[FAIL] {t.__name__}: {e}")
    passed = len(tests) - failed - skipped
    print(f"{passed} passed, {skipped} skipped, {failed} failed "
          f"(of {len(tests)})")
    sys.exit(1 if failed else 0)


if __name__ == "__main__":
    main()

File: bindings/cpp/test_zcio.cpp
// Smoke test for the zcio C++ RAII wrapper. Build:
//   c++ -std=c++17 -I../../include test_zcio.cpp -L../../build -lzcio -o test_cpp
#include <zcio/zcio.hpp>

#include <cassert>
#include <cstdio>
#include <string>
#include <thread>
#include <vector>
#include <unistd.h>

// Derive a per-process port base so rapid re-runs avoid TIME_WAIT collisions
// on fixed ports. Range stays well inside the ephemeral space.
static int port_base() {
    return 39000 + (static_cast<int>(::getpid()) % 2000);
}

static int failures = 0;
#define CHECK(cond) do { if (!(cond)) { \
    std::fprintf(stderr, "FAIL %s:%d: %s\n", __FILE__, __LINE__, #cond); ++failures; } } while (0)

int main() {
    zcio::init();

    // version
    CHECK(!zcio::version().empty());

    // ring
    {
        zcio::Ring r(64);
        CHECK(r.write(std::string_view("hello")) == 5);
        CHECK(r.available() == 5);
        char buf[5];
        CHECK(r.read(buf, 5) == 5);
        CHECK(std::string(buf, 5) == "hello");
    }

    // serializer operator<< / >> over a buffer (all scalar widths)
    {
        unsigned char mem[256];
        {
            auto s = zcio::Serial::buffer(mem, sizeof mem);
            s << int8_t(-1) << uint8_t(250) << int16_t(-2) << uint16_t(60000)
              << int32_t(-7) << uint32_t(7u) << int64_t(-9) << uint64_t(9u)
              << float(2.5f) << double(1.5) << std::string("vibe");
        }
        {
            auto s = zcio::Serial::buffer(mem, sizeof mem);
            int8_t i8 = 0; uint8_t u8 = 0; int16_t i16 = 0; uint16_t u16 = 0;
            int32_t i = 0; uint32_t u = 0; int64_t i64 = 0; uint64_t u64 = 0;
            float f = 0; double d = 0; std::string str;
            s >> i8 >> u8 >> i16 >> u16 >> i >> u >> i64 >> u64 >> f >> d >> str;
            CHECK(i8 == -1); CHECK(u8 == 250);
            CHECK(i16 == -2); CHECK(u16 == 60000);
            CHECK(i == -7); CHECK(u == 7u);
            CHECK(i64 == -9); CHECK(u64 == 9u);
            CHECK(f == 2.5f); CHECK(d == 1.5);
            CHECK(str == "vibe");
        }
    }

    // counting serial
    {
        auto s = zcio::Serial::counting();
        s << int32_t(0) << double(0);
        CHECK(s.write_pos() == 4 + 8);
    }

    // serial positioning + status accessors over a buffer
    {
        unsigned char mem[64];
        auto s = zcio::Serial::buffer(mem, sizeof mem);
        s << int32_t(11) << int32_t(22);
        CHECK(s.write_pos() == 8);
        s.set_read_pos(4);
        int32_t v = 0;
        s >> v;
        CHECK(v == 22);
        CHECK(s.read_pos() == 8);
        CHECK(!s.short_read());
    }

    // bit-array roundtrip over a Ring stream
    {
        zcio::Ring ring(64);
        std::vector<uint8_t> bits = {1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0};
        {
            auto s = zcio::Serial::over(ring.as_stream(), /*bit_stream=*/true);
            s.write_bits(bits, 0, bits.size());
            s.synchronize();
        }
        {
            auto s = zcio::Serial::over(ring.as_stream(), /*bit_stream=*/true);
            std::vector<uint8_t> out;
            s.read_bits(out, 0, bits.size());
            for (size_t i = 0; i < bits.size(); ++i) CHECK(out[i] == bits[i]);
        }
    }

    // stream copy: Ring -> memory Stream, plus available/eof on StreamRef
    {
        zcio::Ring src(64);
        src.write(std::string_view("copy-me"));
        char dstbuf[16] = {};
        auto dst = zcio::memory_stream(dstbuf, sizeof dstbuf);
        zcio::StreamRef srcref = src.as_stream();
        int64_t n = zcio::copy(dst, srcref, 7);
        CHECK(n == 7);
        CHECK(std::string(dstbuf, 7) == "copy-me");
    }

    // tcp loopback (back-compat accept())
    try {
        const int port = port_base() + 0;
        zcio::TcpServer srv(port);
        zcio::TcpClient cli("127.0.0.1", port);
        auto conn = srv.accept(2000);
        CHECK(static_cast<bool>(conn));
        if (conn) {
            cli.stream().write(std::string_view("roundtrip"));
            char buf[9];
            CHECK(conn.read_full(buf, 9) == 9);
            CHECK(std::string(buf, 9) == "roundtrip");
        }
    } catch (const zcio::Error& e) {
        std::printf("[skip] tcp loopback: %s\n", e.what());
    }

    // tcp server multi-client: accept 2, close one
    try {
        const int port = port_base() + 1;
        zcio::TcpServer srv(port);
        zcio::TcpClient c1("127.0.0.1", port);
        auto a1 = srv.accept_conn(2000);
        CHECK(static_cast<bool>(a1));
        zcio::TcpClient c2("127.0.0.1", port);
        auto a2 = srv.accept_conn(2000);
        CHECK(static_cast<bool>(a2));
        CHECK(a1.id != a2.id);
        // exercise wait_* / bytes_available
        c1.stream().write(std::string_view("ping"));
        CHECK(a1.wait_readable(1000) >= 0);
        char b[4];
        CHECK(a1.stream.read_full(b, 4) == 4);
        CHECK(std::string(b, 4) == "ping");
        CHECK(c1.wait_writable(1000) >= 0);
        // close the first client by id
        CHECK(srv.close_client(a1.id) == 0);
    } catch (const zcio::Error& e) {
        std::printf("[skip] tcp multi-client: %s\n", e.what());
    }

    // udp loopback round-trip
    try {
        const int port = port_base() + 2;
        zcio::UdpServer srv(port);
        zcio::UdpClient cli("127.0.0.1", port);
        cli.stream().write(std::string_view("udp-hi"));
        auto pkt = srv.receive(false, 1000000);
        CHECK(static_cast<bool>(pkt));
        if (pkt) {
            char buf[6];
            int64_t got = pkt.read(buf, sizeof buf);
            CHECK(got == 6);
            CHECK(std::string(buf, 6) == "udp-hi");
        }
    } catch (const zcio::Error& e) {
        std::printf("[skip] udp loopback: %s\n", e.what());
    }

    // multicast best-effort loopback (must not fail if nothing arrives)
    {
        try {
            const int port = port_base() + 3;
            zcio::McastReceiver rx("239.255.0.1", port);
            zcio::McastSender tx("239.255.0.1", port);
            tx.stream().write(std::string_view("mcast"));
            char buf[5] = {};
            int64_t got = rx.stream().read(buf, sizeof buf);
            if (got == 5 && std::string(buf, 5) == "mcast")
                std::printf("multicast loopback delivered\n");
            else
                std::printf("[skip] multicast: no datagram in window\n");
        } catch (const zcio::Error& e) {
            std::printf("[skip] multicast unavailable: %s\n", e.what());
        }
    }

    // DNS
    {
        CHECK(zcio::is_ipv4("127.0.0.1"));
        CHECK(!zcio::is_ipv4("not-an-ip"));
        CHECK(zcio::resolve_ipv4("127.0.0.1") == "127.0.0.1");
        std::string lan = zcio::local_ipv4();
        if (!lan.empty()) std::printf("local ipv4: %s\n", lan.c_str());
        else std::printf("[skip] local ipv4 unavailable\n");
        auto a = zcio::dns_query_a("127.0.0.1");
        CHECK(!a.empty());
        CHECK(a[0] == "127.0.0.1");
        // AAAA on a literal is best-effort; just exercise it.
        auto aaaa = zcio::dns_query_aaaa("::1");
        (void)aaaa;
    }

    // HTTP offline: tiny responder on a thread via zcio::TcpServer.
    try {
        const int port = port_base() + 4;
        zcio::TcpServer srv(port);
        std::thread responder([&srv]() {
            auto conn = srv.accept_conn(3000);
            if (conn) {
                char buf[1024];
                conn.stream.read(buf, sizeof buf);  // consume request line/headers
                static const char* resp =
                    "HTTP/1.1 200 OK\r\n"
                    "Content-Length: 5\r\n"
                    "Connection: close\r\n"
                    "\r\n"
                    "hello";
                conn.stream.write(std::string_view(resp));
                conn.stream.flush();
                // http_get's read_all loops until EOF, so close the connection.
                srv.close_client(conn.id);
            }
        });
        auto r = zcio::http_get("http://127.0.0.1:" + std::to_string(port) + "/");
        responder.join();
        CHECK(r.status == 200);
        CHECK(r.body == "hello");
    } catch (const zcio::Error& e) {
        std::printf("[skip] http offline: %s\n", e.what());
    }

    // TLS loopback (skip cleanly if no backend compiled in)
    {
        if (!zcio::tls_available()) {
            std::printf("[skip] TLS backend unavailable\n");
        } else {
            const int port = port_base() + 5;
            try {
                auto sctx = zcio::TlsCtx::server();
                zcio::TcpServer srv(port, sctx);
                std::thread server_thr([&srv]() {
                    auto conn = srv.accept(3000);
                    if (conn) {
                        char b[5] = {};
                        if (conn.read_full(b, 5) == 5)
                            conn.write(std::string_view(b, 5));
                        conn.flush();
                    }
                });
                auto cctx = zcio::TlsCtx::client("127.0.0.1");
                zcio::TcpClient cli("127.0.0.1", port, cctx, /*verify=*/false);
                cli.stream().write(std::string_view("tlsok"));
                cli.stream().flush();
                char rb[5] = {};
                bool ok = cli.stream().read_full(rb, 5) == 5
                          && std::string(rb, 5) == "tlsok";
                server_thr.join();
                CHECK(ok);
                if (ok) std::printf("TLS loopback ok (backend=%s)\n",
                                    zcio::tls_backend_name().c_str());
            } catch (const zcio::Error& e) {
                std::printf("[skip] TLS loopback: %s\n", e.what());
            }
        }
    }

    if (failures == 0) std::printf("all C++ wrapper tests passed\n");
    return failures ? 1 : 0;
}
